Hashing strategy interface and the H2, H3, H4, H5 functions

pdyraga opened the pull request:
~~Depends on #2~~

The goal of this PR is to abstract out ciphersuite hashing strategy and add the missing `H2`, `H3`, `H4`, and `H5` hash functions.

The `Hash` interface abstracts out hash functions implementations specific to the ciphersuite used. This is a strategy design pattern allowing to use FROST with different ciphersuites, like secp256k1 or Jubjub curves.

All `H*` functions but `H2` use the tag defined in FROST. Those functions are internal to the FROST protocol so we will use the FROST tag for a better domain separation. Only `H2` must use BIP340 version given how the BIP340 verification algorithm is specified.

pdyraga committed:
H1 frost function, BIP034 hashing function

This changeset adds H1 function as specified in section 6.5. FROST(secp256k1,
SHA-256) of [FROST]. The [BIP340] hash function is used as expand_message_xmd.
The implementation is copied from @eth-r's prototype. I added tests and
annotated relevant parts of the hash functions code with reference to papers.
The `testutils` library code was borrowed from the `keep-core` repository.

pdyraga committed:
Fixed the value of the contextString

The values was not matching the one explained in the comment above. Now
they do match.

pdyraga committed:
Removed references to HASH-TO-CURVE

We are not using the implementation from HASH-TO-CURVE. We use BIP-340
version of hashing functions.

pdyraga committed:
Abstract out ciphersuite hashing strategy behind the Hash interface

Hash interface abstracts out hash functions implementations specific to the
ciphersuite used. This is a strategy design pattern allowing to use FROST with
different ciphersuites, like secp256k1 or Jubjub curves.

[FROST] requires the use of a cryptographically secure hash function,
generically written as H. Using H, [FROST] introduces distinct domain-separated
hashes, H1, H2, H3, H4, and H5. The details of H1, H2, H3, H4, and H5 vary
based on ciphersuite.

pdyraga committed:
H2 function for BIP-340 hashing interface

H2 is the implementation of H2(m) function from [FROST] implemented in a way
compatible to how [BIP340] hashing functions are specified.

In the interface, we expect at least one message to be provided. Optionally, we
allow for more and they will all be concatenated. Expecting at least one message
makes sense for the hash function signature. We allow for more because it is how
this function will be used in FROST.

pdyraga commented on roast/hash.go:
```diff
@@ -5,22 +5,57 @@ import (
+// based on ciphersuite.
+type Hash interface {
+	H1(m []byte) *big.Int
+	H2(m []byte, ms ...[]byte) *big.Int
```
In the interface, we expect at least one message to be provided. Optionally, we allow for more and they will all be concatenated. Expecting at least one message makes sense for the hash function signature. We allow for more to be provided in the second argument because it is how this function will be used in FROST.

eth-r replied:
Because `H2` has to conform to the specification of whatever FROST variant is being implemented, it needs special consideration. Different protocols have different rules for how to encode points in the signature and its verification. We can use multiple `[]byte`s as arguments if we use another interface (e.g. `Curve`) to define the encoding rules. Otherwise it'll need to be `H2(gc *Point, pk *Point, msg []byte) *big.Int` or something similar.

pdyraga replied:
I was thinking about having a `Cyphersuite` interface we would pass to the FROST protocol. The `Cyphersuite` would specify the particular `Hash`ing strategy but also the `Curve` where we could put the encoding logic.

eth-r replied:
I think something along those lines makes sense.

pdyraga replied:
We could also rename `Hash` interface to `Cyphersuite`. We would keep `H*` and curve operations there. I am not sure which option is better, I'd decide about it later once we have more information.

pdyraga commented on roast/hash.go:
```diff
@@ -5,22 +5,57 @@ import (
-func h1(m []byte) *big.Int {
-	dst := concat(contextString, []byte("rho"))
-	return hashToScalar(dst, m)
+func (b *Bip340Hash) H2(m []byte, ms ...[]byte) *big.Int {
+	// From [FROST], we know the tag should be DST = contextString || "chal".
+	dst := concat(b.contextString(), []byte("chal"))
+	// We use [BIP340]-compatible hashing algorithm and turn the hash into
+	// a scalar, as expected by [FROST] for H2.
+	return b.hashToScalar(dst, concat(m, ms...))
+}
```
Looking at the prototype, the `H2 function` [as implemented in `hash.go`](https://github.com/threshold-network/roast-go/blob/e33d503c97ad970cc4bbf99463e197fc20fde39e/hash.go#L45-L47) is not used at all and the [`BIP340HashChallenge`](https://github.com/threshold-network/roast-go/blob/e33d503c97ad970cc4bbf99463e197fc20fde39e/hash.go#L23C12-L25) is used instead.

BIP-340 uses three hashing functions:
- `hash_BIP0340/nonce`
- `hash_BIP0340/challenge`
- `hash_BIP0340/aux`

FROST(secp256k1, SHA-256) uses five hashing functions:
- `H1` with `DST = contextString || "rho"`
- `H2` with `DST = contextString || "chal"`
- `H3` with `DST = contextString || "nonce"`
- `H4` as `H(contextString || "msg" || m)`
- `H5` as `H(contextString || "com" || m)`

For consistency, I was thinking about using the `FROST-secp256k1-BIP340-v1 || <tag>` everywhere with `<tag> ∈ {"rho", "chal", "nonce", "msg", "com"}`.

`H1`, `H2`, `H3` would turn the hash into scalar using modulo on secp256k1 curve order and return `*big.Int`. `H4` and `H5` would return `[]byte`.

What do you think about it @eth-r?

eth-r replied:
`H2` must match the BIP-340 spec for how the challenge in the BIP-340 Schnorr signature is calculated, so we can't do as proposed because the signatures would be invalid: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#user-content-Design

Everything else is internal to the protocol and we can implement however we wish.

pdyraga replied:
I fixed it in fc77336053858a58d753f7e8616a4bc4e24ad29d.

eth-r reviewed (changes requested):
The BIP-340 spec limits the `H2` function which needs to be accounted for. Everything else seems fine.

pdyraga committed:
Use BIP034 tag for H2 hashing implementation

For H2, we need to use [BIP034] tag because the verification algorithm
from [BIP034] expects this tag to be used:

Let e = int(hash_BIP0340/challenge(bytes(r) || bytes(P) || m)) mod n.

pdyraga committed:
H3, H4, and H5 hash functions

H3, H4, and H5 functions use the tag defined in FROST. Those functions are
internal to the FROST protocol so we will use the FROST tag for a better domain
separation. This is also consistent with H1 and only H2 must use BIP340 version
given how the BIP340 verification algorithm is specified.

pdyraga commented:
I missed changing the base branch before undrafting https://github.com/threshold-network/roast-go/pull/2. As a result, the changes were mistakenly merged to a feature branch and not to `main`. This is not a big deal because this branch was forked from the latest state of #2 so with merging this PR changes from both #2 and #3 will land in `main`.

eth-r reviewed (approved):
Looks good to me.

pdyraga committed:
Improved package documentation for roast
